subsys/power/device_pm.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <string.h>
+#include <soc.h>
+#include <device.h>
+#include "policy/pm_policy.h"
+
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+#define DEV_RPM_ASYNC	(1 << 0)
+
+static void ipm_work_handler(struct k_work *work)
+{
+	struct device_config *cfg = CONTAINER_OF(work,
+					struct device_config, ipm_work);
+	struct device *dev = device_get_binding(cfg->name);
+	int ret;
+
+	k_sem_take(&dev->config->lock, K_FOREVER);
+
+	/* Incase of suspend request, make sure the device is not busy */
+	if (device_busy_check(dev) &&
+			(cfg->ipm_request == DEVICE_PM_SUSPEND_STATE)) {
+		goto out;
+	}
+
+	ret = device_set_power_state(dev, cfg->ipm_request);
```

Comment:
If change of device power state is long, you are blocking while tread executing delayed works in the system.
The same applies for k_sem_take() above.

Reply:
I think we should change the semaphore to spinlock as it would be the right primitive to use here.

On device_set_power_state call, we will have to execute this call at some point, even if it takes longer. How can we address this? eventually, we will have to call this right?

Reply:
I have worked with gyro which needed almost 100ms for wake-up. We have to design the interlace to deal with such devices. IMHO the device_set_power_state() should be asynchronous. Otherwise we will end locking something for 100ms. It will be workqueue, calling thread or whole system, depending where we put the lock.

IMHO Device PM needs an FSM and trace all the transitions. I would like to propose the following scheme:

1. Variables:
- `curent_state`
- `in_transition`
- `target_state`

2. API:
```
state_changed_cb(state)
{
 current_state = state;
 if (target_state != current_state)
    trigger_state_change(target_state, state_changed_cb)
else
  in_transition = false
}
 
set_powert_state(state)
{
  target_state = state
  if (!in_transition && (target_state != current_state)) {
    trigger_state_change(target_state, state_changed_cb)
    in_transition = true;
  } 
}
```

Reply:
@ramakrishnapallala: I will hold with review until we get an idea how to support devices which needs a lot of time to change the state.

Reply:
@pizi-nordic even if we make device_set_power_state() asynchronous the problem will remain the same. Th main issue would the time the device takes to suspend/resume. Until those suspend/resume calls are executed the API's which use these devices will be blocked. In case asynchronous pm_get/put we can add the timeout for the delayed work which would be helpful in case back to back transitions.
If we think the time taken to resume or suspend is not worth power saving or impacts the expected performance then the driver can simply not implement the device idle PM.

Reply:
I've gotta make my usual complaint: what exactly are these functions supposed to do?  Is there design documentation somewhere?  The header only specifies that functions enable, disable, suspend, and resume, with no information about exactly what that's supposed to mean, especially if these operations may become asynchronous or multiple devices are being turned down in sequence.  How do they interact?  What are the possible error codes and what specifically do they mean in this context?

I'm not saying this shouldn't be merged--I haven't been able to look at it closely enough--but I am saying that I have no faith that this API will do what I want when I get around to using LPM (which is very important to me).

Reply:
@pabigot I will update the PR with more documentation. Thanks.

Reply:
> The only thing is it uses the system thread to execute the handler.

And this is my concern. Switching power state of external device connected to SPI or I2C might take hundreds of us. During that time system thread servicing work queue will be blocked. We can mitigate this by allowing registration of custom thread for PM management. What do you think?

Reply:
yeah, that makes a lot of sense. I will add the support and update the PR. Thanks.

Reply:
@pizi-nordic can you point me to an external device that has the behavior you describe?  I'd like to understand the operations involved and the cause of the delay to see if an approach I'm prototyping that gets rid of using threads within sensor drivers might apply.

Reply:
@pabigot: Here you have few examples:
- Any device connected to I2C: For bus speed 100kHz, 8-bit device address, 8-bit register address and 8-bit value written there to put device in idle mode, you will spent at least 270us waiting for I2C transaction to complete. 

- LIS3DH Accelerometer: Change of operating mode takes 1/ODR (ODR - internal clock called "Data Rate" in the datasheet, which could be as low as 1Hz) + 1.6ms. 

- Any device with analog part which needs some time to reach operating point (PLL, Accelerometer, Gyroscope, Temp Sensor and so on).

Reply:
Thanks; that's the sort of thing I was expecting.